pyblazing/pyblazing/apiv2/context.py

```diff
@@ -3236,3 +3214,169 @@ def log(self, query, logs_table_name="bsql_logs"):
+            miniters=1,
+            bar_format=self._get_progress_bar_format(),
+            leave=False,
+        )
+        pbar2 = tqdm(
+            miniters=1, bar_format="Total Batches Processed: {n_fmt}", leave=False
+        )
+
+        pbar.update(thesteps)
+        pbar2.update(batches_completed)
+        last = thesteps
+        last_sum_batches = batches_completed
+        while True:
+            query_complete = graph.query_is_complete()
+            progress = graph.get_progress()
+            pdf = queryProgressAsPandas(progress)
+            batches_completed = pdf["batches_completed"].sum()
+            pdf = pdf.drop(pdf[~pdf.finished].index)
+            thesteps = len(pdf)
+            if last != thesteps:
+                delta = thesteps - last
+                last = thesteps
+                pbar.update(delta)
+                if last_sum_batches != batches_completed:
+                    delt2 = batches_completed - last_sum_batches
+                    pbar2.update(delt2)
+                    last_sum_batches = batches_completed
+            sleep(0.005)
+            if query_complete:
+                break
+        pbar.close()
+        pbar2.close()
+
+    def _check_tqdm(self):
+        tqdm_found = True
+        try:
+            from tqdm import tqdm
+        except ImportError:
+            tqdm_found = False
+            err_msg = "Warning: Could not set the progress bar, please install tqdm"
+            print(err_msg)
+        return tqdm_found
+
+    def _run_progress_bar_distributed(self, ctxToken):
+        from tqdm import tqdm
+
+        ispbarCreated = False
+        pbar = None
+        query_complete = False
+        last = -1
+        themax = -1
+        last_sum_batches = -1
+        while True:
+            dask_futures = []
+            for node in self.nodes:
+                worker = node["worker"]
+                dask_futures.append(
+                    self.dask_client.submit(
+                        getQueryIsComplete, ctxToken, workers=[worker], pure=False
+                    )
+                )
+            workers_is_complete = self.dask_client.gather(dask_futures)
+            query_complete = all(workers_is_complete)  # all workers returned true
+
+            dask_futures = []
+            for node in self.nodes:
+                worker = node["worker"]
+                dask_futures.append(
+                    self.dask_client.submit(
+                        getQueryProgress, ctxToken, workers=[worker], pure=False
+                    )
+                )
+            workers_progress = dask.dataframe.from_delayed(dask_futures).compute()
+            themax = len(workers_progress)
+            pdf = workers_progress
+            batches_completed = pdf["batches_completed"].sum()
+            pdf = pdf.drop(pdf[~pdf.finished].index)
+            thesteps = len(pdf)
+            if not ispbarCreated:
+                ispbarCreated = True
+                pbar = tqdm(
```

Comment:
Please move the initializations of the progress bars to outside the while loop, the way you have it in _run_progress_bar_single_node

Reply:
@williamBlazing sure but I will need to call all the workers in order to get the max value for the progress bar

Reply:
I see. Well, in general, see if you can refactor this a little. If not, its not a big deal. But do please add to the README and meta.yaml about the new requirements